### ISSUE TITLE
Demos of fpc-wayland application.

### ISSUE THREAD

**Issue by fredvs** [andrewd207/fpc-wayland#1]
Hello Andrews,

Many thanks for this precious project.

Just for info, there is a project of Wayland demos using your impressive work.
https://github.com/fredvs/wayland-pascal

![wylandclient_xdg_chessboard](https://github.com/andrewd207/fpc-wayland/assets/3421249/16355d90-d907-46ce-8eb8-a9450531976c)

**Comment by andrewd207** [comment]
@fredvs  How awesome! :D I wrote this a while ago and revisited it at some point. Amazing how it seems more complicated later. I'm glad somebody can use it. I started also to write a completely native interface using only the file handles which was working but I arrived again at a point how to generate the interfaces from the xml files and stopped there.

**Comment by fredvs** [comment]
Ho, Andrew, having some news from you makes me so happy.
I very worried not having heard from you.


About Wayland, hum, it is a jungle there, each "tenor" widgetset wants to impose his own compositor, so not easy to have something "universal".

Also only docs and C demos obsolete, so it is a real total dark jungle.

**Comment by andrewd207** [comment]
I was looking at your code just now and from what I saw you are using the raw "c" style bindings. Did you try the object oriented bindings at all? One of my primary goals was to make it familiar for modern pascal styles to use. I initially started it to use with fpGUI but I was a bit stumped when it came to doing window decorations. There was some talk about a kde protocol to have the system paint the window border but that's where I stopped.

**Comment by fredvs** [comment]
Yes, indeed, I have try to use your object oriented code but did not catch how to use it.
So I mixed part of your code + raw "c" style.

Of course your way is much cleaner but without demo I am lost.

That said, you are welcome to translate the "steps" demos into oop code. ;-)

**Comment by andrewd207** [comment]
I did a very basic example

```
program Project1;

{$mode objfpc}{$H+}

uses
  {$IFDEF UNIX}
  cthreads,
  {$ENDIF}
  Classes, sysutils,
  wayland_protocol, wayland_client_core, xdg_shell_protocol
  { you can add units after this };

type

  { TMyApp }

  TMyApp = class(IWlRegistryListener)
  private
    FDisplay: TWLDisplay;
    FRegistry: TWlRegistry;
    FCompositor: TWlCompositor;
    procedure wl_registry_global(AWlRegistry: TWlRegistry; AName: DWord; AInterface: String; AVersion: DWord);
    procedure wl_registry_global_remove(AWlRegistry: TWlRegistry; AName: DWord);
  public
    constructor Create;
    destructor Destroy; override;
  end;

{ TMyApp }

procedure TMyApp.wl_registry_global(AWlRegistry: TWlRegistry; AName: DWord; AInterface: String; AVersion: DWord);
var
  lProxy: Pwl_proxy;

begin
  WriteLn('Available interface: ', AInterface, '(', AName, ')');
  if AInterface = wl_compositor_interface.name then
  begin
    lProxy := AWlRegistry.Bind(AName, @wl_compositor_interface, wl_compositor_interface.version);
    FCompositor := TWlCompositor.Create(lProxy, True);
  end;



end;

procedure TMyApp.wl_registry_global_remove(AWlRegistry: TWlRegistry;
  AName: DWord);
begin
  WriteLn('Remove interface: ', AName);
end;

constructor TMyApp.Create;
var
  lSurface: TWlSurface;
begin
  FDisplay := TWlDisplay.Connect('') as TWLDisplay;
  if not Assigned(FDisplay) then
  begin
    WriteLn('Failed to open display');
    Exit;
  end;

  FRegistry := FDisplay.GetRegistry();
  FRegistry.AddListener(Self);
  FDisplay.Roundtrip;
  if Assigned(FCompositor) then
  begin
    lSurface := FCompositor.CreateSurface();
    WriteLn('Surface creation successful = ', Assigned(lSurface));
    FreeAndNil(lSurface);
  end;
end;

destructor TMyApp.Destroy;
begin
  FreeAndNil(FCompositor);
  FreeAndNil(FRegistry);
  FreeAndNil(FDisplay);
  inherited;
end;

begin
  with TMyApp.Create do
    Free;
end.

```

**Comment by fredvs** [comment]
Thanks for the demo!
Out-of-the-box as usual with your code.

![fpc-wayland-project1](https://github.com/andrewd207/fpc-wayland/assets/3421249/f78061df-4cdc-4246-a8af-5f5fb56f29a7)

**Comment by fredvs** [comment]
> I was looking at your code just now and from what I saw you are using the raw "c" style bindings. 

Yes, all the "steps" demos are translation of C code given in the WaylandBook.
The original C code (updated) is here, each project has his translation in Pascal code:
https://github.com/fredvs/wayland-pascal/tree/main/src/c

Note that those C demos use inline code that is not accessible via exported methods of the library.
In your wayland_protocol.pas those mehods are added using the equivalent of those inlined methods.

**Comment by andrewd207** [comment]
I have made some changes to the generator and also added a Twl_fixed type to help with the 24.8 fixed decimal type. Also I made some demo's for the TProxyClass wrappers and added them. Hopefully nothing has broken :)

**Comment by fredvs** [comment]
Hello Andrew.

Many thanks for this.
I will test it asap.
Write you later.

Fre;D

**Comment by fredvs** [comment]
Re-hello.

I have followed this gist:
https://gist.github.com/probonopd/9feb7c20257af5dd915e3a9f2d1f2277

There are interesting  things, each clan has his valuable arguments.

**Comment by fredvs** [comment]
Re-re hello.

I have to confess that one of my goal is to enable Wayland for MSEgui too, like for fpGUI.
https://github.com/mse-org/mseide-msegui

So, after the discovery of "pure Wayland" it would be great to have a kind of translator from X11 to Wayland.
But maybe it is not possible, all must be re-write from scratch.

**Comment by fredvs** [comment]
Re-....-hello.

About fpGUi.
There are some issues fixed, could it be possible to commit those fixes?
The same for the pull-requests.

Maybe fpGUI is on standby state, so we have to wait a few more time.

**Comment by andrewd207** [comment]
> Re-re hello.
> 
> I have to confess that one of my goal is to enable Wayland for MSEgui too, like for fpGUI. https://github.com/mse-org/mseide-msegui
> 
> So, after the discovery of "pure Wayland" it would be great to have a kind of translator from X11 to Wayland. But maybe it is not possible, all must be re-write from scratch.

As far as I know there is no Wayland implementation for fpGUI. I started making one using my bindings but it's very incomplete. The most complicated part is implementing a title bar and borders of a window inside the toolkit and invisible to whatever app someone uses.

**Comment by fredvs** [comment]
I did try with your branch:
https://github.com/graemeg/fpGUI/tree/wayland

>The most complicated part is implementing a title bar and borders of a window inside the toolkit and invisible to whatever app someone uses.

Yes, it is the same for the WaylandBook, the steps to do decorations of windows and using compositor are "TODO" steps...

**Comment by andrewd207** [comment]
Wow I completely forgot I commited that :rofl: 
if you look at this file, it has no dependencies on fpGui but abstracts wayland to be a bit easier to use.
https://github.com/graemeg/fpGUI/blob/wayland/src/corelib/wayland/fpg_wayland_classes.pas

**Comment by fredvs** [comment]
>https://github.com/graemeg/fpGUI/blob/wayland/src/corelib/wayland/fpg_wayland_classes.pas

Good meat!
Xxxl thanks.

Fre;D

**Comment by fredvs** [comment]
About the fpGUI UIDesigner for Wayland.
How do you see it at begining: keep the UIDesigner of X11 for Wayland projects and for the compilation, kind of cross-compilation from X11 to Wayland or re-create a "pure Wayland" UIDesigner with from designer and widgets palette "pure Wayland"?

Maybe, for beginning, keeping the X11 UIDesigner working in a XWayland session would be easier.

But if fpGui for Wayland is working, of course, compiling UIDesigner for Wayland should be better.

But all this is hard work, even GTK and Qt suffer to make their widgetset Wayland compatible.
